### ISSUE TITLE
CanvasControl with CustomDevice and DeviceLost

### ISSUE THREAD

**Issue by benstevens48** [Microsoft/Win2D#615]
If I set CustomDevice on a CanvasControl, then try to test DeviceLost handing by disabling my display adapter, the CanvasControl is cleared, but does not subsequently raise any Draw events and also does not raise the CreateResources event with the Reason being NewDevice. The DeviceLost event on my custom device is also never called. Therefore there doesn't seem to be any way to see that the device needs to be re-created. Is there some other way to do this? If not, can I suggest that either an API is added to enable this, or CreateResources with Reason of NewDevice should be fired even if using a CustomDevice.

**Comment by benstevens48** [comment]
Actually, I think that the most logical thing that should happen is that after the device is lost, the CanvasControl should call RaiseDeviceLost on the CustomDevice. (I assume that after a device is lost, when the canvas control decides it needs to redraw, it encounters an error on creating the drawing session or something and checks to see if it's a device lost error, and at that point if there is no custom device presumably it creates a new device and invokes CreateResources with Reason NoDevice, and I think if there is a CustomDevice it should instead call RaiseDeviceLost on the custom device).

**Comment by shawnhar** [comment]
I don't see this behavior.  The first time the control is redrawn after the device is lost, it will raise the device lost event.

Is it possible that your control is never getting redrawn?  Device lost is not detected until you try to render something, so if it's just displaying static content nothing will happen until something changes to invalidate the control.

**Comment by benstevens48** [comment]
Hi Shawn, 
Are you sure you have set the CustomDevice? I'm definitely calling Invalidate on the CanvasControl after the device has been lost, but it neither raises a Draw event on the CanvasControl nor a DeviceLost event on the CustomDevice.

**Comment by benstevens48** [comment]
Here is a very simple app which demonstrates it - <https://1drv.ms/u/s!AnQ-_3An97TMgqkhUs1_yuUpRKK_Kg>

By running the app, then disabling your display adapter and looking at the debug text output you can see that DeviceLost is not called when using a custom device (but works fine without a custom device). Tap on the page to explicitly invalidate the canvas control.

**Comment by shawnhar** [comment]
Thanks for the repro.  Debugging this led me to the relevant code inside Win2D, from which I was able to dig up some ancient design emails - and it turns out this behavior is actually by design :-)

The rationale is that if CanvasControl automatically created its device, it should be responsible for handling lost devices (raising the event, creating a new device and redrawing as needed, etc).   But if the app has provided a custom device, that is presumably because they are doing something unusual that Win2D doesn't know about.  Maybe that device has special requirements in some way, or was obtained through interop with some other API.   In this case the CanvasControl has no ability to recreate the device, and it can't be sure what the intended model for lost device handling is, so it just does nothing and waits for the caller to provide a new working device.

Therefore, it's up to you to detect and recover from lost devices if you are using a custom device.

**Comment by benstevens48** [comment]
Ok, but that means there's no way for me to recover from a lost device since I'm using the canvas control to do the drawing. As far as I can see the only way would be to set up a timer and periodically attempt to do something which would fail with a device lost error if the device was lost. This doesn't seem ideal, and it would presumably be quite easy for the canvas control to raise the device lost event when it tried to do some drawing. If you don't plan to allow this, can you suggest how I can detect a device lost event? If I use the timer method above, then what is the cheapest operation I can do that will fail with a device lost event and doesn't have any side-effects? Thanks.

**Comment by damyanp** [comment]
Can you use [ID3D11Device4::RegisterDeviceRemovedEvent](https://msdn.microsoft.com/en-us/library/windows/desktop/mt589890(v=vs.85).aspx)?

**Comment by shawnhar** [comment]
Can I ask why you are using a custom device in the first place?

If you want Win2D to help you with recovering from device lost, you should just let Win2D create and manage the device for you.

The main rationale for custom device support is to allow combining Win2D with other frameworks that already create devices and have their own system for managing those.  The assumption being that if you are using such another framework, that will already have its own system for handling the lost device case.

**Comment by benstevens48** [comment]
@shawnhar  Ok, I understand about the other frameworks.

I'm not sure I 100% need to create the device myself but my reason is as follows. I'm using Win2D to render some images since it has nice interpolation options with immediate zoom. Due to the logical structure of the app the images are loaded first and the canvas is created on demand (and then reused) to display the image (this also allows the app to start loading the image immediately and not wait for the canvas create resources event), and also each image may be rendered to one of two canvases, so I thought I'd better explicitly make sure everything was using the same device. Or is it safe to rely on everything using the same shared device?

**Comment by benstevens48** [comment]
@damyanp That looks like it would do what I want. Any chance it could be exposed in the C# API. For the time being I have implemented the following workaround.

```
    //class variable CanvasDevice device has already been created and 
    //DisplayInformation.DisplayContentsInvalidated += DisplayInformation_DisplayContentsInvalidated registered

    private void DisplayInformation_DisplayContentsInvalidated(DisplayInformation sender, object args) {
            var localDevice = device;
            if (localDevice != null) {
                try {
                    var bitmap = CanvasBitmap.CreateFromColors(device, new Color[] { Colors.Black }, 1, 1);
                    bitmap.Dispose();
                } catch (Exception ex) {
                    if (localDevice.IsDeviceLost(ex.HResult)) {
                        localDevice.RaiseDeviceLost();
                    }
                }
            }
        }
```

This seems reasonable as DisplayInformation.DisplayContentsInvalidated doesn't seem to be called very often. Or is there a better method I can call other than CanvasBitmap.CreateFromColors to get the device lost error?

**Comment by shawnhar** [comment]
> Due to the logical structure of the app the images are loaded first and the canvas is created on demand (and then reused) to display the image (this also allows the app to start loading the image immediately and not wait for the canvas create resources event), and also each image may be rendered to one of two canvases, so I thought I'd better explicitly make sure everything was using the same device.

This is exactly what GetSharedDevice (and CanvasControl.UsedSharedDevice) are for.  There's no need to create the device yourself in order to have it shared across multiple controls.

Be aware, though, that if you load content outside of the CreateResources event handler, you won't get the benefit of all the work that CanvasControl does on your behalf to make lost device handling robust.  It's certainly possible to do all of this yourself (see https://microsoft.github.io/Win2D/html/LoadingResourcesOutsideCreateResources.htm) but quite subtle and takes a while to get right (at least, it took us a while to get the version built into CanvasControl right :-)

**Comment by benstevens48** [comment]
Ok, thanks. I wasn't sure that GetSharedDevice was guaranteed to return the same instance provided the device had not been lost - I thought maybe it was primarily a performance optimisation rather than to guarantee things were using the same device instance. 

In the end I have actually decided to use CanvasVirtualImageSource directly instead. Given that I'm already creating resources and handling device lost using my own code, it didn't seem like any extra effort to use it instead of a canvas control. The only extra thing I had to do was call Resize if the container I'm fitting it to resizes or the DPI changes. It seems to work fine.
